server/src/lib/server.rs

```diff
@@ -371,43 +437,132 @@ async fn instance_serial(
+        ));
+    }
+    if request
+        .headers()
+        .get(header::SEC_WEBSOCKET_VERSION)
+        .map(|v| v.as_bytes())
+        != Some(b"13")
+    {
+        return Err(HttpError::for_bad_request(
+            None,
+            "missing or invalid websocket version".to_string(),
+        ));
+    }
+    let accept_key = request
+        .headers()
+        .get(header::SEC_WEBSOCKET_KEY)
+        .map(|hv| hv.as_bytes())
+        .map(|key| handshake::derive_accept_key(key))
+        .ok_or_else(|| {
+            HttpError::for_bad_request(
+                None,
+                "missing websocket key".to_string(),
+            )
         })?;
+
+    let (detach_sender, detach_recv) = oneshot::channel();
+    context.serial_detach = Some(detach_sender);
+
+    let upgrade_fut = upgrade::on(&mut *request);
+    let serial = context.serial.clone();
+    let ws_log = rqctx.log.new(o!());
+    let err_log = ws_log.clone();
+    tokio::spawn(
+        async move {
+            let upgraded = upgrade_fut.await?;
+            let ws_stream =
+                WebSocketStream::from_raw_socket(upgraded, Role::Server, None)
+                    .await;
+            instance_serial_task(detach_recv, serial, ws_stream, ws_log).await
+        }
+        .inspect_err(move |err| error!(err_log, "Serial Task Failed: {}", err)),
+    );
+
+    Ok(Response::builder()
+        .status(StatusCode::SWITCHING_PROTOCOLS)
+        .header(header::CONNECTION, "Upgrade")
+        .header(header::UPGRADE, "websocket")
+        .header(header::SEC_WEBSOCKET_ACCEPT, accept_key)
+        .body(Body::empty())?)
+}
+
+#[endpoint {
+    method = PUT,
+    path = "/instances/{instance_id}/serial/detach",
+}]
+async fn instance_serial_detach(
+    rqctx: Arc<RequestContext<Context>>,
+    path_params: Path<api::InstancePathParams>,
+) -> Result<HttpResponseUpdatedNoContent, HttpError> {
+    let mut context = rqctx.context().context.lock().await;
+
+    let context = context.as_mut().ok_or_else(|| {
+        HttpError::for_internal_error(
+            "Server not initialized (no instance)".to_string(),
+        )
+    })?;
+    if path_params.into_inner().instance_id != context.properties.id {
+        return Err(HttpError::for_internal_error(
+            "UUID mismatch (path did not match struct)".to_string(),
+        ));
     }
-    let response = api::InstanceSerialResponse { bytes: output[..n].to_vec() };
-    Ok(HttpResponseOk(response))
+
+    let serial_detach =
+        context.serial_detach.take().filter(|s| !s.is_closed()).ok_or_else(
+            || {
+                HttpError::for_bad_request(
+                    None,
+                    "serial console already detached".to_string(),
+                )
+            },
+        )?;
+
+    serial_detach.send(()).map_err(|_| {
+        HttpError::for_internal_error(
+            "couldn't send detach message to serial task".to_string(),
+        )
+    })?;
```

Comment:
So this "send" doesn't actually necessarily mean the serial console is detached fully, right?

Theoretically, couldn't we:

- (instance_serial_task tokio task): write to serial console
- (instance_serial_detach HTTP handler): `take` the `context.serial_detach` object, invoke the `send` oneshot
- (instance_serial HTTP handler): someone *new* tries connecting to the serial console, sees `context.serial_detach` is none.
- (OLD instance_serial_task tokio task): keep writing to serial console (`select` didn't happen to pick 'detach').

I would argue that we should hold onto the tokio_task handle and join during this detach, to ensure that the old serial console has exited fully before we return from `detach`.

Reply:
Good catch. This should be handled now.